x-pack/plugins/enterprise_search/server/lib/indices/pipelines/ml_inference/pipeline_processors/delete_ml_inference_pipeline.ts

```diff
@@ -18,6 +21,12 @@ export const deleteMlInferencePipeline = async (
 ) => {
   let response: DeleteMlInferencePipelineResponse = {};
 
+  // Check if the pipeline is in use in a different index's managed pipeline
+  const pipelineIsInUse = await isPipelineInUse(pipelineName, indexName, client);
+  if (pipelineIsInUse) {
+    throw new Error(ErrorCode.PIPELINE_IS_IN_USE);
+  }
+
   try {
     response = await detachMlInferencePipeline(indexName, pipelineName, client);
```

Comment:
I'm not following how we could change `response` to a constant without altering the logic.
The `detachMlInferencePipeline()` call is within a `try/catch` block for the reason of suppressing a potential 404 (graceful handling of an edge case).
Then `deletePipeline()` is called outside the `try/catch`, which mutates `response` before returning it.

Could you elaborate how we could have a constant that's visible outside the `try/catch` but initialized inside it?

Reply:
You can move the block where you call `client.ingest.deletePipeline` inside the same try block. Does that make sense to you?

Reply:
It makes sense; what I'm saying is reorganizing the code like that would alter the logic of the flow. Whether or not `client.ingest.deletePipeline()` throws an error on a HTTP 404, today we don't have that in the scope of the graceful handling block in `catch`. With the reorg we would be introducing a potential bug (swallowing a 404 error from that call that is meant to bubble up) only for the sake of eliminating the mutable variable.

Even if we _do_ know that `deletePipeline()` doesn't throw that kind of error today, it could in the future, which makes this potential bug even worse.

What I'm suggesting instead is extracting the detaching + error handling logic into a local function:

```js
const detachPipeline = (
  indexName: string,
  pipelineName: string,
  client: ElasticsearchClient
): Promise<DeleteMlInferencePipelineResponse> => {
  try {
    return detachMlInferencePipeline(indexName, pipelineName, client);
  } catch (error) {
    // only suppress Not Found error
    if (error.meta?.statusCode !== 404) {
      throw error;
    }

    return Promise.resolve({});
  }
}

// call in deleteMlInferencePipeline():
const response: DeleteMlInferencePipelineResponse = await detachPipeline(indexName, pipelineName, client);
```

This way we focus the error handling to just `detachMlInferencePipeline()` and we also make `response` a constant. How does that sound?

(We don't have the same problem with the other variable - `let pipelines` - because the processing code that goes within `try/catch` is local and doesn't throw that specific error. So I'll refactor that as you suggested.)

Reply:
Ah, I get what you're saying now. Yes, that should be fine. An alternative approach would be to wrap the detachMlInferencePipeline call in an inner try catch (you can just nest try/catches), but that feels a little ugly.

sidenote: you don't need the Promise.resolve() there if you just mark that function as async. You also don't need to explicitly type the response when you call it, TypeScript should correctly infer that from function usage.